pkg/proxy/proxy.go

```diff
@@ -264,51 +266,221 @@ func (p *Proxy) health(ctx echo.Context) error {
+	}
+
+	return nil
+}
+
+// validateSpace checks whether the Space exists.
+func (p *Proxy) validateSpace(workspaceName string) error {
+	if _, err := p.app.InformerService().GetSpace(workspaceName); err != nil {
+		// log the actual error but do not return it so that it doesn't reveal information about a space that may not belong to the requestor
+		log.Error(nil, err, "unable to get target cluster for workspace "+workspaceName)
+		return fmt.Errorf("the requested space is not available")
+	}
+	return nil
+}
+
+// validateUserAccess checks whether the user is Approved, if they are not an error is returned.
+// If public-viewer is enabled, user validation is skipped.
+func (p *Proxy) validateUserAccess(ctx echo.Context, userID, username string) error {
+	// skip if public-viewer is enabled: read-only operations on community workspaces are always permitted.
+	if context.IsPublicViewerEnabled(ctx) {
+		return nil
+	}
+
+	// retrieve the UserSignup for the requesting user.
+	//
+	// UserSignup complete status is not checked, since it might cause the proxy blocking the request
+	// and returning an error when quick transitions from ready to provisioning are happening.
+	userSignup, err := p.app.SignupService().GetSignupFromInformer(nil, userID, username, false)
+	if err != nil {
+		return err
+	}
+
+	// if the UserSignup is nil or has NOT the CompliantUsername set,
+	// it means that MUR was NOT created and useraccount is NOT provisioned yet
+	// TODO(@filariow): should we also check if the user is Banned?
+	if userSignup == nil || userSignup.CompliantUsername == "" {
+		cause := errs.New("user is not provisioned (yet)")
+		log.Error(nil, cause, fmt.Sprintf("signup object: %+v", userSignup))
+		return cause
+	}
+	return nil
+}
+
+// getClusterAccess retrieves the access to the cluster hosting the requested workspace,
+// if the user has access to it.
+// Access can be either direct (an SpaceBinding linking the user to the workspace exists)
+// or community (a SpaceBinding linking the PublicViewer user to the workspace exists).
+func (p *Proxy) getClusterAccess(ctx echo.Context, userID, username, workspaceName, proxyPluginName string, workspaces []toolchainv1alpha1.Workspace) (*access.ClusterAccess, error) {
+	// retrieve by name the workspace from the list of workspaces the user has access to.
+	w := findWorkspaceByName(workspaceName, workspaces)
+	if w == nil {
+		return nil, crterrors.NewInternalError(errs.New("unable to get target cluster"), "workspace not found")
+	}
+
+	// retrieve the requesting user's UserSignup
+	signup, err := p.spaceLister.GetSignupFunc(nil, userID, username, false)
+	if err != nil {
+		return nil, crterrors.NewInternalError(errs.New("unable to get target cluster"), "user not found")
+	}
+
+	// if PublicViewer is enabled and user has no direct access to the workspace, proceed as PublicViewer
+	publicViewerEnabled := context.IsPublicViewerEnabled(ctx)
+	if publicViewerEnabled && !hasDirectAccess(signup, w) {
```

Comment:
maybe I'm making some confusion here - but shouldn't there be a binding for `KubesawAuthenticatedUsername` before giving public viewer access to this workspace ? It seems like we only check if the public viewer feature is enabled.

Reply:
that's indeed a nice to have to avoid flooding the target cluster! let me work on this

Reply:
I double-checked, and the access for public viewer is already guaranteed from `SpaceLister.GetUserWorkspace`.
It returns information about the Workspace only if the user has direct access or if PublicViewer support is enabled and PublicViewer user has access to the Workspace. So here we can assume that if we have data about the workspace and user has no direct access than the PublicViewer user has direct access to it. Hence, adding this check is not required. 

At the same time it is very easy to add this check, it could increase readability and robustness of the code without impacting much the performances. Do you suggest adding it?

```diff
-	if publicViewerEnabled && !hasDirectAccess(signup, w) {
+	if publicViewerEnabled && !hasDirectAccess(signup, w) && hasDirectAccess(kubesawAuthenticatedSignup, w) {
```

Reply:
oh gotcha! yeah it wasn't clear at the first, I couldn't remember all the checks that were performed before this if statement , adding it explicitly might help indeed , if this doesn't affect performances. Thanks for checking this!

Reply:
Yeah, it's easy to forget some of this check. Even I had to double check. Also, the fact that this PR is so big is not helping. I'll try to break it in multiple PRs